middleware/service_broker_rate_limiter.rb

```diff
@@ -0,0 +1,101 @@
+module CloudFoundry
+  module Middleware
+    class ServiceBrokerRequestCounter
+      include Singleton
+
+      def initialize
+        @mutex = Mutex.new
+        @data = {}
+      end
+
+      def can_make_request?(user_guid, limit)
+        @mutex.synchronize do
+          request_count = @data.fetch(user_guid, 0)
+          return false if request_count + 1 > limit
+
+          @data[user_guid] = request_count + 1
+          true
+        end
+      end
+
+      def decrement(user_guid)
+        @mutex.synchronize do
+          @data[user_guid] = @data[user_guid] - 1
+        end
+      end
+    end
+
+    class ServiceBrokerRateLimiter
+      def initialize(app, logger:, concurrent_limit:)
+        @app                               = app
+        @logger                            = logger
+        @concurrent_limit = concurrent_limit
+        @request_counter = ServiceBrokerRequestCounter.instance
+      end
+
+      def call(env)
+        request = ActionDispatch::Request.new(env)
+        user_guid = env['cf.user_guid']
+
+        unless skip_rate_limiting?(env, request)
+          return too_many_requests!(env) unless @request_counter.can_make_request?(user_guid, @concurrent_limit)
+
+          begin
+            return @app.call(env)
+          rescue => e
+            raise e
+          ensure
```

Comment:
Leaking the counter could be really bad as its hard to notice over a lot of instances as its in memory and people could get sporadic 429s that are hard to trace. Are we sure that in any error situation the CC code will decrement that counter ? I lack the knowledge if there is a way that the middleware is skipped or the ensure is not called. Like e.g aborted connections or similar low level errors that make answering impossible.
It looks like that the Middleware is called with `@app.call(env)` so in any case the ensure should run ?